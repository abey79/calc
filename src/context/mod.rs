pub mod ast;
pub mod checked_ast;
pub mod source;
pub mod token_stream;
