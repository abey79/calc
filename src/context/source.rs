pub struct Source(String);

impl Source {
    pub fn new(source: String) -> Self {
        Self(source)
    }

    pub fn source(&self) -> &str {
        &self.0
    }
}
