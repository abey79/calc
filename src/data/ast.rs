//! This module contains the AST data structures.
//!
//! Throughout the code, the "Node" terminology is used to refer to elements of the AST. AST nodes
//! are assigned a unique `NodeId` when created. Factory functions are also provided to make it easy
//! to build AST nodes.

use crate::data::identified::{new_id, Identified};
use crate::data::token::TokenKind;
use std::fmt;

// =================================================================================================
// DATA STRUCTURES

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct NodeId(usize);

impl NodeId {
    pub fn new() -> Self {
        Self(new_id())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOpKind {
    Pos,
    Neg,
}

pub type UnaryOp = Identified<UnaryOpKind, NodeId>;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

pub type BinOp = Identified<BinOpKind, NodeId>;

pub type VarName = Identified<String, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind {
    Variable(VarName),
    BinOp {
        op: BinOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    UnaryOp {
        op: UnaryOp,
        operand: Box<Expr>,
    },
    Integer(i32),
    Float(f64),
}

pub type Expr = Identified<ExprKind, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub enum StmtKind {
    Assign { name: VarName, value: Expr },
    Print { expr: Expr },
    Expr { expr: Expr },
}

pub type Stmt = Identified<StmtKind, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub struct BlockKind {
    pub stmts: Vec<Stmt>,
}

pub type Block = Identified<BlockKind, NodeId>;

// =================================================================================================
// FACTORIES

impl From<&TokenKind> for UnaryOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Pos,
            TokenKind::Minus => Self::Neg,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl UnaryOp {
    pub fn new(kind: impl Into<UnaryOpKind>) -> Self {
        Self {
            kind: kind.into(),
            id: NodeId::new(),
        }
    }
}

impl From<&TokenKind> for BinOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Add,
            TokenKind::Minus => Self::Sub,
            TokenKind::Star => Self::Mul,
            TokenKind::Slash => Self::Div,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl BinOp {
    pub fn new(kind: impl Into<BinOpKind>) -> Self {
        Self {
            kind: kind.into(),
            id: NodeId::new(),
        }
    }
}

impl VarName {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            kind: name.into(),
            id: NodeId::new(),
        }
    }
}

impl Expr {
    pub fn variable(name: impl Into<VarName>) -> Self {
        Self {
            kind: ExprKind::Variable(name.into()),
            id: NodeId::new(),
        }
    }

    pub fn bin_op(op: BinOp, left: Expr, right: Expr) -> Self {
        Self {
            kind: ExprKind::BinOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            },
            id: NodeId::new(),
        }
    }

    pub fn unary_op(op: UnaryOp, operand: Expr) -> Self {
        Self {
            kind: ExprKind::UnaryOp {
                op,
                operand: Box::new(operand),
            },
            id: NodeId::new(),
        }
    }

    pub fn integer(value: i32) -> Self {
        Self {
            kind: ExprKind::Integer(value),
            id: NodeId::new(),
        }
    }

    pub fn float(value: f64) -> Self {
        Self {
            kind: ExprKind::Float(value),
            id: NodeId::new(),
        }
    }
}

impl Stmt {
    pub fn assign(name: impl Into<VarName>, value: Expr) -> Self {
        Self {
            kind: StmtKind::Assign {
                name: name.into(),
                value,
            },
            id: NodeId::new(),
        }
    }

    pub fn print(expr: Expr) -> Self {
        Self {
            kind: StmtKind::Print { expr },
            id: NodeId::new(),
        }
    }

    pub fn expr(expr: Expr) -> Self {
        Self {
            kind: StmtKind::Expr { expr },
            id: NodeId::new(),
        }
    }
}

impl Block {
    pub fn new(stmts: Vec<Stmt>) -> Self {
        Self {
            kind: BlockKind { stmts },
            id: NodeId::new(),
        }
    }
}

// =================================================================================================
// UTILITY TRAITS

impl fmt::Display for NodeId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl fmt::Display for BinOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
        }
    }
}

impl fmt::Display for UnaryOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Pos => write!(f, "+"),
        }
    }
}

impl From<&str> for VarName {
    fn from(s: &str) -> Self {
        Self {
            kind: s.to_string(),
            id: NodeId::new(),
        }
    }
}
