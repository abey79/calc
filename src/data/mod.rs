pub mod ast;
pub mod identified;
pub mod span;
pub mod token;
