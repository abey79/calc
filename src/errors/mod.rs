pub mod error_context;

use error_context::ErrorContext;
use thiserror::Error;

/// Errors generated by the tokenizer
#[derive(Error, Debug, PartialEq)]
pub enum SyntaxError {
    #[error("unexpected end of file")]
    UnexpectedEndOfFile,

    #[error("unexpected character '{0}'")]
    UnexpectedCharacter(char),
}

/// Errors generated by the tokenizer
#[derive(Error, Debug, PartialEq)]
pub enum TokenizerError {
    #[error("{1}Syntax error:{0}")]
    SyntaxError(SyntaxError, ErrorContext),
}
