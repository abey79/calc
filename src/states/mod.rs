//! This module models states (as in "state machine") between the compilation steps.
//!
//! The data associated with each state is factored in "context" structure for reusability (e.g. the
//! [`TokenStream`] context ("state data") is used in both the [`TokenizedState`] and the
//! [`ParsedState`]).

pub mod input_state;
pub mod parsed_state;
pub mod tokenized_state;

pub use input_state::*;
pub use parsed_state::*;
pub use tokenized_state::*;
