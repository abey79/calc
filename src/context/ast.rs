use crate::data::ast::Stmt;
use crate::data::meta::Meta;
use crate::data::span::Span;
use crate::data::token_span::TokSpan;
use crate::pipeline;
use std::fmt;
use std::fmt::{Debug, Display, Write};

#[derive(Debug, Default)]
pub struct Ast<M: Debug + Display> {
    stmts: Vec<Stmt<M>>,
}

impl<M: Debug + Display> Ast<M> {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[Stmt<M>] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<Stmt<M>> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: Stmt<M>) {
        self.stmts.push(stmt);
    }

    pub fn format<W: fmt::Write>(&self, w: &mut W) -> Result<String, fmt::Error> {
        pipeline::formatter::format(self, w)
    }
}

impl Ast<TokSpan> {
    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in self.stmts() {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Provide convenient methods for AST nodes parametrized by a token span.
impl<K> Meta<K, TokSpan> {
    pub fn tok_span(&self) -> TokSpan {
        self.meta.clone()
    }

    pub fn span(&self) -> Span {
        self.meta.span()
    }
}
