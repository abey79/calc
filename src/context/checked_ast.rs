//! State after the type checker.
//!
//! In this state, all the AST nodes are decorated with type info.

use crate::data::ast::{Expr, Stmt};
use crate::data::token_span::TokSpan;
use std::fmt;
use std::fmt::Write;

pub type CheckedStmt = Stmt<TypeInfo>;
pub type CheckedExpr = Expr<TypeInfo>;

#[derive(Debug, Default)]
pub struct CheckedAst {
    stmts: Vec<CheckedStmt>,
}

impl CheckedAst {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[CheckedStmt] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<CheckedStmt> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: CheckedStmt) {
        self.stmts.push(stmt);
    }

    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in &self.stmts {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Note:
// This is a massive simplification of what a real type structure would be. In a realistic case,
// `CheckedAst` would contain a vector of types (including user defined ones). Then, `CheckedInfo`
// would contain a ref-counted pointer to one of the types.

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Type {
    Stmt, // stmt only
    Integer,
    Float,
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Type::*;
        match self {
            Stmt => write!(f, "stmt"),
            Integer => write!(f, "int"),
            Float => write!(f, "float"),
        }
    }
}

/// AST meta-data after type checking.
#[derive(Debug)]
pub struct TypeInfo {
    /// type of the node
    pub type_: Type,

    /// token span of the node (typically copied over from the parsed state)
    pub tok_span: TokSpan,
}

impl TypeInfo {
    pub fn new(type_: Type, tok_span: TokSpan) -> Self {
        Self { type_, tok_span }
    }
}

impl fmt::Display for TypeInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.type_)
    }
}
