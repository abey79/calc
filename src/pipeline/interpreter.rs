use crate::context::checked_ast::{CheckedExpr, CheckedStmt};
use crate::data::ast::{BinOpKind, ExprKind, StmtKind, UnaryOpKind};
use crate::errors::{InterpreterError, Spanned, SyntaxError, TypeError};
use crate::states::CheckedState;
use std::collections::HashMap;
use std::fmt;
use std::fmt::Write;

type Result<T> = std::result::Result<T, InterpreterError>;

pub(crate) fn interpret<W: Write>(input: &CheckedState, writer: &mut W) -> Result<()> {
    let mut interpreter = Interpreter::new(input, writer);
    interpreter.run()
}

#[derive(Debug, Clone)]
pub enum Value {
    Int(i32),
    Float(f64),
    Tuple(Vec<Value>),
}

impl Value {
    fn bin_op(&self, op: &BinOpKind, other: &Value) -> Option<Value> {
        match (self, other) {
            (Self::Int(i1), Self::Int(i2)) => match op {
                BinOpKind::Add => Some(Self::Int(i1 + i2)),
                BinOpKind::Sub => Some(Self::Int(i1 - i2)),
                BinOpKind::Mul => Some(Self::Int(i1 * i2)),
                BinOpKind::Div => Some(Self::Int(i1 / i2)),
            },
            (Self::Float(f1), Self::Float(f2)) => match op {
                BinOpKind::Add => Some(Self::Float(f1 + f2)),
                BinOpKind::Sub => Some(Self::Float(f1 - f2)),
                BinOpKind::Mul => Some(Self::Float(f1 * f2)),
                BinOpKind::Div => Some(Self::Float(f1 / f2)),
            },
            (Self::Tuple(t1), Self::Tuple(t2)) => {
                // tuple addition and subtraction are element-wise
                if t1.len() != t2.len() || !matches!(op, BinOpKind::Add | BinOpKind::Sub) {
                    None
                } else {
                    let mut res = Vec::new();
                    for (v1, v2) in t1.iter().zip(t2) {
                        res.push(v1.bin_op(op, v2)?);
                    }
                    Some(Self::Tuple(res))
                }
            }
            (Self::Tuple(t), Self::Int(i)) | (Self::Int(i), Self::Tuple(t)) => {
                // tuple multiplication and division is scalar multiplication
                if !matches!(op, BinOpKind::Mul | BinOpKind::Div) {
                    None
                } else {
                    let mut res = Vec::new();
                    for v in t {
                        res.push(v.bin_op(op, &Self::Int(*i))?);
                    }
                    Some(Self::Tuple(res))
                }
            }
            (Self::Tuple(t), Self::Float(fl)) | (Self::Float(fl), Self::Tuple(t)) => {
                // tuple multiplication and division is scalar multiplication
                if !matches!(op, BinOpKind::Mul | BinOpKind::Div) {
                    None
                } else {
                    let mut res = Vec::new();
                    for v in t {
                        res.push(v.bin_op(op, &Self::Float(*fl))?);
                    }
                    Some(Self::Tuple(res))
                }
            }
            _ => None,
        }
    }

    fn unary_op(&self, op: &UnaryOpKind) -> Value {
        match self {
            Self::Int(i) => match op {
                UnaryOpKind::Pos => Self::Int(*i),
                UnaryOpKind::Neg => Self::Int(-*i),
            },
            Self::Float(f) => match op {
                UnaryOpKind::Pos => Self::Float(*f),
                UnaryOpKind::Neg => Self::Float(-*f),
            },
            Self::Tuple(values) => match op {
                UnaryOpKind::Pos => Self::Tuple(values.clone()),
                UnaryOpKind::Neg => Self::Tuple(values.iter().map(|v| v.unary_op(op)).collect()),
            },
        }
    }
}

impl fmt::Display for Value {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Value::Int(i) => i.fmt(f),
            Value::Float(fl) => write!(f, "{:?}", fl),
            Value::Tuple(values) => {
                write!(f, "(")?;
                for (i, value) in values.iter().enumerate() {
                    if i != 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{}", value)?;
                }
                write!(f, ")")
            }
        }
    }
}

struct Interpreter<'a, W: Write> {
    input: &'a CheckedState,
    writer: &'a mut W,

    // state
    vars: HashMap<String, Value>,
}

impl<'a, W: Write> Interpreter<'a, W> {
    fn new(input: &'a CheckedState, writer: &'a mut W) -> Self {
        Self {
            input,
            writer,
            vars: HashMap::new(),
        }
    }

    fn run(&mut self) -> Result<()> {
        for stmt in self.input.ast.stmts() {
            self.run_stmt(stmt)?;
        }
        Ok(())
    }

    fn run_stmt(&mut self, stmt: &CheckedStmt) -> Result<()> {
        match &stmt.kind {
            StmtKind::Assign { name, value } => {
                let value = self.run_expr(value)?;
                self.vars.insert(name.kind.clone(), value);
            }
            StmtKind::Print { expr } => {
                let value = self.run_expr(expr)?;
                writeln!(self.writer, "{}", value)?;
            }
            StmtKind::Expr { expr } => {
                // somewhat pointless as no side effects as possible in this language
                let _ = self.run_expr(expr)?;
            }
        }
        Ok(())
    }

    fn run_expr(&mut self, expr: &CheckedExpr) -> Result<Value> {
        match &expr.kind {
            ExprKind::Variable(name) => {
                let value = self.vars.get::<String>(name.as_ref()).ok_or_else(|| {
                    InterpreterError::SyntaxError(
                        SyntaxError::UnknownVariable(name.to_string()),
                        name.to_error(&self.input.source),
                    )
                })?;
                Ok(value.clone())
            }
            ExprKind::BinOp { op, left, right } => {
                let left_val = self.run_expr(left)?;
                let right_val = self.run_expr(right)?;
                let value = left_val.bin_op(&op.kind, &right_val).ok_or_else(|| {
                    // this should never happen as the type checker should have caught this
                    InterpreterError::TypeError(
                        TypeError::MismatchedTypesForBinaryOp(
                            left.meta.type_.clone(),
                            right.meta.type_.clone(),
                        ),
                        op.to_error(&self.input.source),
                    )
                })?;
                Ok(value)
            }
            ExprKind::UnaryOp { op, operand } => {
                let value = self.run_expr(operand)?;
                Ok(value.unary_op(&op.kind))
            }
            ExprKind::Tuple(exprs) => {
                let mut values = Vec::new();
                for expr in exprs {
                    values.push(self.run_expr(expr)?);
                }
                Ok(Value::Tuple(values))
            }
            ExprKind::Integer(i) => Ok(Value::Int(*i)),
            ExprKind::Float(fl) => Ok(Value::Float(*fl)),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::pipeline::checker::check;
    use crate::pipeline::interpreter::interpret;
    use crate::pipeline::parser::parse;
    use crate::pipeline::tokenizer::tokenize;
    use crate::states::InputState;

    #[test]
    fn test_interpreter() {
        let input = InputState::from(
            r###"
                a = (1.3 + 3.2) * 45.1;
                print a;
                b = a * 3.2;
                print b;
                print 1 + 2 * 3;
            "###,
        );
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();

        let mut output = String::new();
        interpret(&checked, &mut output).unwrap();

        insta::assert_snapshot!(output);
    }
}
