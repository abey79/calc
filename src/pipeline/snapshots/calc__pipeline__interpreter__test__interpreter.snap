---
source: src/pipeline/interpreter.rs
expression: output
---
"202.95000000000002\n649.44\n7\n"
